engine/src/flutter/shell/platform/darwin/ios/framework/Source/TextRange.swift

```diff
@@ -0,0 +1,113 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import UIKit
+
+/// A UITextPosition subclass that represents a location in the text, in
+/// a Flutter text field.
+@objc(FlutterTextPosition)
+final class TextPosition: UITextPosition, NSCopying {
+  // offset is exposed as `range` in an extension on UITextPosition.
+  fileprivate let offset: UInt
+  @objc let affinity: UITextStorageDirection
+
+  init(index: UInt, affinity: UITextStorageDirection = .forward) {
+    self.offset = index
+    self.affinity = affinity
+  }
+
+  override func isEqual(_ object: Any?) -> Bool {
+    guard let other = object as? TextPosition else {
+      return false
+    }
+    // Affinity is only visual.
+    return offset == other.offset
+  }
+
+  func copy(with zone: NSZone? = nil) -> Any {
+    return TextPosition(index: offset, affinity: affinity)
+  }
+
+  override var description: String {
+    "TextPosition(\(index))"
+  }
+}
+
+/// A UITextRange subclass that represents a text range within the text, in
+/// a Flutter text field.
+///
+/// Similar to the dart:ui TextRange class, this class represents a right open
+/// interval consists of two `TextPosition`s.
+@objc(FlutterTextRange)
+final class TextRange: UITextRange, NSCopying {
+  // nsRange is exposed as `range` in an extension on UITextPosition.
+  fileprivate let nsRange: NSRange
+
+  init(NSRange range: NSRange) {
+    self.nsRange = range
+  }
+
+  func copy(with zone: NSZone? = nil) -> Any {
+    TextRange(NSRange: nsRange)
+  }
+
+  override var start: TextPosition {
+    TextPosition(index: UInt(nsRange.lowerBound))
+  }
+
+  override var end: TextPosition {
+    TextPosition(index: UInt(nsRange.upperBound))
+  }
+
+  override var isEmpty: Bool { range.length == 0 }
+
+  override func isEqual(_ object: Any?) -> Bool {
+    guard let other = object as? TextRange else {
+      return false
+    }
+    return range == other.nsRange
+  }
+
+  override var description: String {
+    "TextRange(\(range))"
+  }
+}
+
+// MARK: Class methods for initializing in Objective-C
+
+@available(swift, obsoleted: 1.0)
```

Comment:
why do we need this

Reply:
This is extension is for objective-c only. If you're calling from swift you can just use the initializer.